### ISSUE TITLE
Dynamic require of "async_hooks" is not supported

### ISSUE THREAD

**Issue by AeShevch** [highlight/highlight#6747]
Greetings.

I requested a problem yesterday regarding BatchSpanProcessorBase - [#6741.](https://github.com/highlight/highlight/issues/6741)

@deltaepsilon fixed it, I updated to 3.4.1, but now I'm getting a new error:

```
error - Error: Dynamic require of "async_hooks" is not supported
    at node_modules/@highlight-run/node/dist/index.mjs:12:9
    at node_modules/@opentelemetry/context-async-hooks/build/src/AsyncHooksContextManager.js (node_modules/@highlight-run/node/dist/index.mjs:17388:22)
    at __require2 (node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at node_modules/@opentelemetry/context-async-hooks/build/src/index.js (node_modules/@highlight-run/node/dist/index.mjs:17511:19)
    at __require2 (node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at node_modules/@opentelemetry/sdk-trace-node/build/src/NodeTracerProvider.js (node_modules/@highlight-run/node/dist/index.mjs:24089:33)
    at __require2 (node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at node_modules/@opentelemetry/sdk-trace-node/build/src/index.js (node_modules/@highlight-run/node/dist/index.mjs:24144:19)
    at __require2 (node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at node_modules/@opentelemetry/sdk-node/build/src/index.js (node_modules/@highlight-run/node/dist/index.mjs:90905:21)
    at __require2 (node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at node_modules/@highlight-run/node/dist/index.mjs:120160:31
    at ModuleJob.run (node:internal/modules/esm/module_job:193:25)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:530:24)
    at async importModuleDynamicallyWrapper (node:internal/vm/module:438:15)
```

To Reproduce

1. npm install --save @highlight-run/node ("@highlight-run/node": "3.4.0",)
2. import { H } from '@highlight-run/node';
3. H.init({ projectID: '***' });
4. Getting the error

**Comment by deltaepsilon** [comment]
I have another patch fix coming out later today.

Keep the error reports coming!!

**Comment by Vadman97** [comment]
4.4.2 is now released - https://www.npmjs.com/package/@highlight-run/next/v/4.4.2

**Comment by AeShevch** [comment]
@Vadman97 @deltaepsilon Still facing the issue on highlight-run/node 3.4.2

**Comment by AeShevch** [comment]
> > @Vadman97 @deltaepsilon Still facing the same issue on highlight-run/node 3.4.2
> 
> Thanks for the heads up!
> 
> Are you having the problem locally or in a deployment? What version of Node are you using?

Locally, nodejs 16.19.0. nextjs 13.5.3.

I just write "import { H } from '@highlight-run/next'" and immediately get an error.

**Comment by GermanJablo** [comment]
Same problem here.
- pages router + trpc
- npm
- removed .next and lock + fresh install
- There is no need to debug in Vercel, the error happens locally.

related: https://discord.com/channels/1026884757667188757/1155921015369629696/1157008643561099285

**Comment by deltaepsilon** [comment]
I got a reproduction!!! It's the `withHighlightConfig` import for `next.config.js`.

Get rid of `withHighlightConfig` for now, and I'll get it figured out. The issue appears to be that `next.config.js` is getting run in the Edge runtime, which is currently incompatible with `withHighlightConfig`.

I'll sort this out early next week.

**Comment by AeShevch** [comment]
@deltaepsilon 
Thanks for your reply! 
The problem occurs when using `H.init` in the Routes API (pages/api). 
I don't use `@highlight-run/next`, only `@highlight-run/node`, so I don't use `withHighlightConfig`. 
I did it this way because I don't need full integration with nextjs. 
I just want to send logs like - request-response on several api routes. 

Can I achieve this using `@highlight-run/next`?

I'm using `npm`

**Comment by deltaepsilon** [comment]
Ah!! 

That's excellent detail. Thank you. I'll test that use case separately.

**Comment by gabrielalmeida** [comment]
Seeing the same while using `@highlight-run/next` without `withHighlightConfig` but wrapping Routes API with `withHighlight` HOC from `@highlight-run/next/server` as well:

```ts
import { Highlight } from '@highlight-run/next/server';

export const withHighlight = Highlight({
  projectID: 'XXX',
});
```

Stacktrace:
```
  automatically enabled Fast Refresh for 1 custom loader
 ○ compiling /instrumentation ...
 ✓ Compiled /instrumentation in 5.4s (168 modules)
Error: An error occurred while loading instrumentation hook: Dynamic require of "async_hooks" is not supported
// -> it seems it still uses highlight/node though    
at file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:12:9
    at node_modules/@opentelemetry/context-async-hooks/build/src/AsyncHooksContextManager.js (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:17388:22)
    at __require2 (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at node_modules/@opentelemetry/context-async-hooks/build/src/index.js (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:17511:19)
    at __require2 (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at node_modules/@opentelemetry/sdk-trace-node/build/src/NodeTracerProvider.js (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:24089:33)
    at __require2 (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at node_modules/@opentelemetry/sdk-trace-node/build/src/index.js (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:24144:19)
    at __require2 (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at node_modules/@opentelemetry/sdk-node/build/src/index.js (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:90905:21)
    at __require2 (file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:18:51)
    at file:///workspaces/zapper-dashboard/node_modules/.pnpm/@highlight-run+node@3.4.2/node_modules/@highlight-run/node/dist/index.mjs:120160:31
    at ModuleJob.run (node:internal/modules/esm/module_job:194:25)
```

```sh
$ node -v
v18.17.0

$ pnpm -v
8.7.5
```

**Comment by Honason** [comment]
Hello! @gabrielalmeida I was having the same error. After updating to @highlight-run/next ^5.0.0, and replacing import paths in both `instrumentation.js` and `next.config.js` to `@highlight-run/next/config`, I see no errors.

Because of the release notes I only replaced the import of `withHighlightConfig`, but only after updating the import of `registerHighlight`, the errors are solved. Maybe you are in the same situation?

**Comment by Honason** [comment]
@deltaepsilon I see you fixed this issue for nextjs.config.js by moving ``

**Comment by deltaepsilon** [comment]
> @deltaepsilon I see you fixed this issue for nextjs.config.js by moving `withHighlightConfig` to `@highlight-run/next/config`.
> 
> I think the same problem happens when importing `registerHighlight` in instrumentation.js. Can it be solved with the same solution?
> 
> ```js
> const {registerHighlight} = await import('@highlight-run/next/server');
> ```

If you check out the new docs, I figured out a way to NOT use instrumentation.ts any longer.

We're still exporting registerHighlight, but I'm hoping to remove it. It was working sometimes... but the API endpoint wrappers have proven more reliable for now.

**Comment by Honason** [comment]
@deltaepsilon Thanks for the info! Endpoint wrappers seem like always more reliable solution. I liked the simplicity of using instrumentation as well though :-)

**Comment by deltaepsilon** [comment]
I wanted instrumentation to work so badly. But reliability wins every time.

It's possible to do auto-wrapping with Webpack, but it's months of work and potentially brittle.

**Comment by spicybackend** [comment]
I'm upgrading `highlight-run/next` from 4.3.x to 5.0.x today and coming across this issue in unit tests for a lot of API endpoints in the pages directory. I'm using Vitest, but assuming the same will occur with Jest and other test frameworks.

To work around the issue I'm mocking out the wrappers entirely in my test setup file. `jest.mock` should work similarly.
```ts
vi.mock("@highlight-run/next/server", async () => ({
  AppRouterHighlight: () => (fn: never) => fn,
  EdgeHighlight: () => (fn: never) => fn,
  PageRouterHighlight: () => (fn: never) => fn,
}));
```

I've ditched the OpenTelemetry setup and started to wrap each API endpoint individually -- it's a bit more boilerplate but I agree it's more reliable this way, particularly while the Next.js instrumentation hook config is still experimental. Along with removing `registerHighlight`, will the dependency on the `@opentelemetry/api` package also be dropped, or is this being used for more under the hood?

**Comment by Vadman97** [comment]
Dug into this extensively today; apologies for the delay on our end. I've pinpointed the async_hooks errors down to our ESM bundling which is not compatible with opentelemetry SDKs. Putting up a fix that we'll release soon after some more testing! https://github.com/highlight/highlight/pull/6917

**Comment by Vadman97** [comment]
Hi all, this is now released with [@highlight-run/node 3.5.0](https://www.npmjs.com/package/@highlight-run/node?activeTab=versions) and [@highlight-run/next 7.0.1](https://www.npmjs.com/package/@highlight-run/next/v/7.0.1)
Please give it a try!
